Modules/_decimal/_decimal.c

```diff
@@ -3411,6 +3364,48 @@ dict_get_item_string(PyObject *dict, const char *key, PyObject **valueobj, const
     return 0;
 }
 
+/*
+ * Fallback _pydecimal formatting for new format specifiers that mpdecimal does
+ * not yet support. As documented, libmpdec follows the PEP-3101 format language:
+ * https://www.bytereef.org/mpdecimal/doc/libmpdec/assign-convert.html#to-string
+ */
+static PyObject *
+pydec_format(PyObject *dec, PyObject *fmt, decimal_state *state)
+{
+    PyObject *result;
+    PyObject *pydec;
+    PyObject *u;
+
+    if (state->PyDecimal == NULL) {
+        state->PyDecimal = _PyImport_GetModuleAttrString("_pydecimal", "Decimal");
+        if (state->PyDecimal == NULL) {
+            return NULL;
+        }
+    }
+
+    u = dec_str(dec);
+    if (u == NULL) {
+        return NULL;
+    }
+
+    pydec = PyObject_CallOneArg(state->PyDecimal, u);
```

Comment:
> > Is it okay that the PyDecimal instance can have a different context?
> 
> This is an excellent observation! It is the one thing that cannot be caught by deccheck, because the _decimal and _pydecimal contexts are always in sync. _pydecimal.__format__ needs rounding and capitals.
> 
> Maybe the easiest thing would be to add another private method to _pydecimal, something like:
> 
> ```python
> def fallback_format(self, specifier, rounding, caps):
>     # make a new context with rounding and caps
>     # return __format__ with that context
> ```

(trying to move this to a thread at relevant part of the code)

Is the problem here, at the invocation of `Decimal` constructor?

I'm not sure how to call it correctly, or why it isn't enough to adopt the existing context.

Reply:
The decimal constructor is fine. It can only raise on an invalid input (which cannot happen here, since the output of `tosci` is valid.

The following is very quick and dirty and possibly full of mistakes. It is just to show the idea.

In _pydecimal, something like this:

```--- cpython/Lib/_pydecimal.py   2024-02-10 13:48:19.119618369 +0000
+++ XXX/Lib/_pydecimal.py       2024-02-10 22:12:17.513196343 +0000
@@ -3841,6 +3841,10 @@
         # of the formatting to the _format_number function
         return _format_number(adjusted_sign, intpart, fracpart, exp, spec)
 
+    def fallback_format(self, specifier, rounding, caps):
+        c = Context(rounding=rounding, capitals=caps)
+        return self.__format__(specifier, c)
+
 def _dec_from_triple(sign, coefficient, exponent, special=False):
     """Create a decimal instance directly, without any validation,
     normalization (e.g. removal of leading zeros) or argument
```

In _decimal, something like this:

```
--- cpython/Modules/_decimal/_decimal.c 2024-02-10 19:40:48.125845778 +0000
+++ XXX/Modules/_decimal/_decimal.c     2024-02-10 22:17:33.494205488 +0000
@@ -3370,10 +3370,12 @@
  * https://www.bytereef.org/mpdecimal/doc/libmpdec/assign-convert.html#to-string
  */
 static PyObject *
-pydec_format(PyObject *dec, PyObject *fmt, decimal_state *state)
+pydec_format(PyObject *dec, PyObject *context, PyObject *fmt, decimal_state *state)
 {
     PyObject *result;
     PyObject *pydec;
+    PyObject *rounding;
+    PyObject *caps;
     PyObject *u;
 
     if (state->PyDecimal == NULL) {
@@ -3394,7 +3396,12 @@
         return NULL;
     }
 
-    result = PyObject_Format(pydec, fmt);
+    rounding = context_getround(context, NULL);
+    caps = context_getcapitals(context, NULL);
+
+    result = PyObject_CallMethod(pydec, "fallback_format", "(OOO)", fmt, rounding, caps);
+    Py_DECREF(rounding);
+    Py_DECREF(caps);
     Py_DECREF(pydec);
 
     if (result == NULL && PyErr_ExceptionMatches(PyExc_ValueError)) {
@@ -3466,7 +3473,7 @@
             PyMem_Free(fmt);
         }
 
-        return pydec_format(dec, fmtarg, state);
+        return pydec_format(dec, context, fmtarg, state);
     }
 
     if (replace_fillchar) {
```

Reply:
added with unit tests